app/jobs/maintenance_tasks/task_job.rb

```diff
@@ -94,5 +99,31 @@ def on_error(error)
       @run.persist_error(error)
       MaintenanceTasks.error_handler.call(error)
     end
+
+    def dynamic_collection?(collection)
+      return false unless collection.respond_to?(:call)
+
+      # Objects like Proc, Proc (lambda) and Method's .call method actually
+      # forward arguments to the code they wrap. In those cases we want to check
+      # the parameters of the wrapped code, not of the call method (which
+      # accepts any params).
+      parameters = if collection.respond_to?(:parameters)
+        collection
+      else
+        collection.method(:call)
+      end.parameters
+
+      return true if parameters.any? do |parameter|
+        type = parameter.first
+        (type == :key || type == :keyreq) && parameter.last == :cursor
+      end
+
+      return true if parameters.any? do |parameter|
```

Comment:
Why we are iterating over the same collection twice? The content of those loops are very similar, so we can appply boolean logic to avoid this kind of duplication. That way we also remove those early returns and the `false` in the end of the method.

Reply:
I had originally kept them separate because I figured the conditional might get confusing. I've combined them, and opted not to extract any helper methods for now. Let me know what you think.